File: Day-04-Two-Sum/Solution.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        if(nums.empty())
            return nums;
        
        // hashMap[value] = index
        unordered_map<int, int> hashMap;
        vector<int> solution;
        for(int i = 0; i != nums.size(); ++i) {
            hashMap.insert(make_pair(nums[i], i));
        }
        
        for(int j = 0; j != nums.size(); ++j) {
            int remainder = target - nums[j];
            unordered_map<int, int>::iterator it = hashMap.find(remainder);
            if(it != hashMap.end() && it->second != j) {
                if(j > it->second) {
                    solution.push_back(it->second);
                    solution.push_back(j);
                } else {
                    solution.push_back(j);
                    solution.push_back(it->second);
                }                    
                return solution;
            }
        }
        return solution;
    }
};

File: Day-04-Two-Sum/Solution.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        if(nums.length == 0)
            return nums;
        
        int[] solution = new int[2];
        
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        
        for(int j = 0; j < nums.length; j++) {
            int remainder = target - nums[j];
            if(map.containsKey(remainder) && map.get(remainder) != j) {
                if(map.get(remainder) > j) {
                    solution[0] = j;
                    solution[1] = map.get(remainder);
                } else {
                    solution[0] = map.get(remainder);
                    solution[1] = j;
                }
                
                return solution;
            }
        }
        
        return solution;
    }
}
